extraConfigSriov.py

```diff
@@ -38,6 +40,45 @@ def _sno_make_deploy(
     client.oc("create namespace openshift-sriov-network-operator")
     client.oc("label ns --overwrite openshift-sriov-network-operator " "pod-security.kubernetes.io/enforce=privileged " "pod-security.kubernetes.io/enforce-version=v1.24 " "security.openshift.io/scc.podSecurityLabelSync=false")
 
+    ocp_version = version
+    if ocp_version is not None:
+        v, number_of_subs_made = re.subn("^(4[.][0-9]+).*$", "\\1", ocp_version)
+        if number_of_subs_made == 1:
+            ocp_version = v
+        else:
+            ocp_version = None
+
+    for ekey in (
+        "NETWORK_RESOURCES_INJECTOR_IMAGE",
+        "OVS_CNI_IMAGE",
+        "SRIOV_CNI_IMAGE",
+        "SRIOV_DEVICE_PLUGIN_IMAGE",
+        "SRIOV_INFINIBAND_CNI_IMAGE",
+        "SRIOV_NETWORK_CONFIG_DAEMON_IMAGE",
+        "SRIOV_NETWORK_OPERATOR_IMAGE",
+        "SRIOV_NETWORK_WEBHOOK_IMAGE",
+    ):
+        # The caller can set the environment variable for deployment.
+        x = os.environ.get(ekey)
+        if x is None and ocp_version is not None:
+            # If they are unset, we fallback to the default.
+            #
+            # These images were pushed using contrib/scripts/build-images.sh.
+            # You can use the script, to build your own images and push to
+            # your own container registry.
+            x = {
+                'SRIOV_INFINIBAND_CNI_IMAGE': f'quay.io/thaller/ib-sriov-cni:{ocp_version}',
+                'SRIOV_CNI_IMAGE': f'quay.io/thaller/sriov-cni:{ocp_version}',
+                "OVS_CNI_IMAGE": None,
+                "SRIOV_DEVICE_PLUGIN_IMAGE": None,
+                "NETWORK_RESOURCES_INJECTOR_IMAGE": f'quay.io/thaller/sriov-dp-admission-controller:{ocp_version}',
+                "SRIOV_NETWORK_CONFIG_DAEMON_IMAGE": f'quay.io/thaller/sriov-network-config-daemon:{ocp_version}',
+                "SRIOV_NETWORK_WEBHOOK_IMAGE": f'quay.io/thaller/sriov-network-webhook:{ocp_version}',
+                "SRIOV_NETWORK_OPERATOR_IMAGE": f'quay.io/thaller/sriov-network-operator:{ocp_version}',
+            }.get(ekey)
+        if x is not None:
+            deploy_env[ekey] = x
```

Comment:
defaultdict?

Reply:
how should that be used here?

Reply:
First, create a dict with these values (call this defaults):

                'SRIOV_INFINIBAND_CNI_IMAGE': f'quay.io/thaller/ib-sriov-cni:{ocp_version}',
                'SRIOV_CNI_IMAGE': f'quay.io/thaller/sriov-cni:{ocp_version}',
                "OVS_CNI_IMAGE": None,
                "SRIOV_DEVICE_PLUGIN_IMAGE": None,
                "NETWORK_RESOURCES_INJECTOR_IMAGE": f'quay.io/thaller/sriov-dp-admission-controller:{ocp_version}',
                "SRIOV_NETWORK_CONFIG_DAEMON_IMAGE": f'quay.io/thaller/sriov-network-config-daemon:{ocp_version}',
                "SRIOV_NETWORK_WEBHOOK_IMAGE": f'quay.io/thaller/sriov-network-webhook:{ocp_version}',
                "SRIOV_NETWORK_OPERATOR_IMAGE": f'quay.io/thaller/sriov-network-operator:{ocp_version}',

Then, We use

deploy_env = {key:  os.environ.get(key, default) for key, default in defaults.items()}


Could that work?

Reply:
I reworked the patch to be more like this. In particular, the environment variables (like "SRIOV_NETWORK_WEBHOOK_IMAGE") are only mentioned once.

`{key: os.environ.get(key, default) for key, default in defaults.items()}` does not work without further changes, because the `None` for `OVS_CNI_IMAGE` needs to be filtered out.

Reply:
```
if version is None:
    # panic out, we need a version to continue
```
we should panic, if-and-only-if the required environment variables are not set when calling `make deploy`. As there are multiple ways how the environment variables get set, that check is easiest done at the end.

```
deploy_env = {key: os.environ.get(key, default) for key, default in default_base_images.items()}
```
for one, `deploy_env` must be merged with the full environment (that also contains KUBECONFIG). Also `default_base_images` contains `{ "OVS_CNI_IMAGE": None }`. That variable doesn't need to be set, because `make deploy` uses a suitable default. It also has on default on quay.io/thaller. For those reasons, the code is still different than this.

Reply:
can we set the image we need in default_base_images for OVS_CNI_IMAGE?

Then, we can have:

```
to_add = {key: os.environ.get(key, default) for key, default in default_base_images.items()}
deploy_env = {**env.copy(), **to_add}
```

Reply:
> can we set the image we need in default_base_images for OVS_CNI_IMAGE?

 I changed now to set `"OVS_CNI_IMAGE": "quay.io/kubevirt/ovs-cni-plugin:latest"`. The difference is that the other entries in `default_base_images` should use the `f":{ocp_version}"` tag, while this one uses `":latest"`. To differentiate that(*), it needs some extra logic. That is the reason, why this is still not a two-liner with dictionary comprehension.

Reply:
Ok, at this point, the whole code looks like this for me:


```

def default_base_images(version: str):
    # These images were pushed using the script mentioned in
    # https://github.com/bn222/cluster-deployment-automation/pull/194. You can
    # use the script, to build your own images and push to your own container
    # registry.
    return {
        "SRIOV_INFINIBAND_CNI_IMAGE": f"quay.io/thaller/ib-sriov-cni:{version}",
        "SRIOV_CNI_IMAGE": f"quay.io/thaller/sriov-cni:{version}",
        "OVS_CNI_IMAGE": "quay.io/kubevirt/ovs-cni-plugin:latest",
        "SRIOV_DEVICE_PLUGIN_IMAGE": f"quay.io/thaller/sriov-network-device-plugin:{version}",
        "NETWORK_RESOURCES_INJECTOR_IMAGE": f"quay.io/thaller/sriov-dp-admission-controller:{version}",
        "SRIOV_NETWORK_CONFIG_DAEMON_IMAGE": f"quay.io/thaller/sriov-network-config-daemon:{version}",
        "SRIOV_NETWORK_WEBHOOK_IMAGE": f"quay.io/thaller/sriov-network-webhook:{version}",
        "SRIOV_NETWORK_OPERATOR_IMAGE": f"quay.io/thaller/sriov-network-operator:{version}",
    }

...

if version is not None:
    version = get_ocp_version()
if version is None:
    # panic out, we need a version to continue
to_add = {key: os.environ.get(key, default) for key, default in default_base_images().items()}
deploy_env = {**env.copy(), **to_add}
```

Am I missing something?

Reply:
if you run CDA with:
```
SRIOV_INFINIBAND_CNI_IMAGE=...
SRIOV_CNI_IMAGE=...
SRIOV_DEVICE_PLUGIN_IMAGE=...
NETWORK_RESOURCES_INJECTOR_IMAGE=...
```
and in `cluster.yaml` have `version=4.17.0-nightly` and `sriov_network_operator_local=True`, then your patch would error out. But there is no reason for that, all images would be correctly set, and non of the `quay.io/thaller` images (which don't have a `4.17` tag) are used.


Or likewise, note that the images at `quay.io/thaller` are necessarily upstream builds. If you want to test something closer to what's in openshift, you would build your own `DOWNSTREAM=1` images (using the script). In that case, you would set all 7 (of the 8) environment variables. There is no need to error out based on not having tags in `quay.io/thaller`, because those images are not used.

Reply:
can't we just add 4.17 now too because that's availabe? Make sure we have all the images for all the versions that are availabe?

Reply:
More specifically, I would like to error out if there is a version that we didn't add yet to CDA. This will force us to add it when it's available.